Gordon360/Services/RecIM/SeriesService.cs

```diff
@@ -0,0 +1,583 @@
+                    TeamStanding = _context.SeriesTeam
+                    .Where(st => st.SeriesID == s.ID)
+                    .Select(st => new TeamRecordViewModel
+                    {
+                        ID = st.ID,
+                        Name = _context.Team
+                                .FirstOrDefault(t => t.ID == st.TeamID)
+                                .Name,
+                        Win = st.Win,
+                        Loss = st.Loss,
+                        //Tie = _context.SeriesTeam
+                        //        .Where(total => total.TeamID == st.TeamID && total.SeriesID == s.ID)
+                        //        .Count() - st.Win - (st.Loss ?? 0)
+                    }).OrderByDescending(st => st.Win).AsEnumerable()
+                });
+            return series;
+        }
+        public SeriesExtendedViewModel GetSeriesByID(int seriesID)
+        {
+            return GetSeries().FirstOrDefault(s => s.ID == seriesID);
+        }
+        public async Task<SeriesViewModel> PostSeriesAsync(SeriesUploadViewModel newSeries, int? referenceSeriesID)
+        {
+            //if activity has no start date
+            var activity = _context.Activity.FirstOrDefault(a => a.ID == newSeries.ActivityID);
+            if (activity.StartDate is null) activity.StartDate = newSeries.StartDate;
+
+            // activity will inherit new end date if series ends after activity ends, OR if activity end date is null
+            if (activity.EndDate is null || activity.EndDate < newSeries.EndDate) activity.EndDate = newSeries.EndDate;
+
+            var series = new Series
+            {
+                Name = newSeries.Name,
+                StartDate = newSeries.StartDate,
+                EndDate = newSeries.EndDate,
+                ActivityID = newSeries.ActivityID,
+                TypeID = newSeries.TypeID,
+                StatusID = 1, //default unconfirmed series
+                ScheduleID = newSeries.ScheduleID ?? 0 //updated when admin is ready to set up the schedule
+            };
+            await _context.Series.AddAsync(series);
+            await _context.SaveChangesAsync();
+
+            IEnumerable<int> teams = new List<int>();
+            if (referenceSeriesID is null)
+            {
+                teams = _context.Team
+                        .Where(t => t.ActivityID == series.ActivityID)
+                        .Select(t => t.ID)
+                        .AsEnumerable();
+            }
+            else
+            {
+                teams = _context.SeriesTeam
+                                .Where(st => st.SeriesID == referenceSeriesID)
+                                .OrderByDescending(st => st.Win)
+                                .Select(st => st.TeamID);
+            }
+            if (newSeries.NumberOfTeamsAdmitted is not null)
+            {
+                teams = teams.Take(newSeries.NumberOfTeamsAdmitted ?? 0);//will never be null but 0 is to silence error
```

Comment:
If it will never be null, why is it nullable?

Reply:
I'll clarify the comment, it'll never be null in that context because there is already a for is not null. However, Take does not allow for a nullable number:
![image](https://user-images.githubusercontent.com/78386128/222860637-c7e37dd5-cad3-45df-88e9-fc0b049a2f9a.png)

Reply:
@amos-cha there are a couple of better ways of handling this.

One of them is to override the error with the [`!` (null-forgiving) operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving), which essentially tells the compiler "I know better than you, ignore the nullability here". That would be correct in this case because we do know for a fact in this branch that `newSeries.NumberOfTeamsAdmitted` isn't `null`.

However, there is an even better way, which is to use pattern matching:
```suggestion
            if (newSeries.NumberOfTeamsAdmitted is int numberOfTeams)
            {
                teams = teams.Take(numberOfTeams);
```
The key difference is that we're catching if `NumberofTeamsAdmitted` is a (non-nullable) `int`, and if so, we're assigning it to the variable `numberOfTeams`. This way, we know that `numberOfTeams` isn't null because otherwise the pattern wouldn't match, and the compiler is happy because `numberOfTeams` is not a nullable type.